// @c4312/blake3-internal/esm/hash-instance.js
import { defaultHashLength, inputToArray } from './hash-fn.js';
import { maxHashBytes } from './hash-reader.js';
export class WasmHasher {
    /** @inhernal */
    constructor(wasmModule, hash) {
        this.wasmModule = wasmModule;
        this.hash = hash;
        this.tookReader = false;
    }
    /** @inheritdoc */
    update(data) {
        // reuse of the hasher is allowed after digesting, but we don't want to
        // change any readers that were already taken. So clone the hash if the
        // user ends up doing this.
        if (this.tookReader) {
            this.hash = this.hash.clone();
            this.tookReader = false;
        }
        const arr = inputToArray(data);
        const saddr = this.hash.scratch.grow(arr.byteLength);
        const step = this.hash.scratch.size;
        for (let i = 0; i < arr.byteLength; i += step) {
            this.wasmModule.HEAPU8.set(arr.subarray(i, Math.min(arr.length, i + step)), saddr);
            this.hash.update(saddr, arr.byteLength);
        }
        return this;
    }
    /** @inheritdoc */
    digest({ length = defaultHashLength } = {}) {
        const out = this.alloc(length);
        const saddr = this.hash.scratch.grow(length);
        const step = this.hash.scratch.size;
        for (let i = 0; i < length; i += step) {
            const n = Math.min(length - i, step);
            this.hash.read(0n, saddr, n);
            out.set(this.wasmModule.HEAPU8.subarray(saddr, saddr + n), i);
        }
        return out;
    }
    /** @inheritdoc */
    reader() {
        const hash = this.hash;
        this.tookReader = true;
        let position = 0n;
        const reader = {
            get position() {
                return position;
            },
            set position(value) {
                if (value > maxHashBytes || value < 0n) {
                    throw new RangeError(`Hash reader position must be within [0, ${maxHashBytes}]`);
                }
                position = value;
            },
            readInto: (target) => {
                const remaining = Number(maxHashBytes - position);
                const length = remaining > target.byteLength ? target.byteLength : remaining;
                const saddr = this.hash.scratch.grow(length);
                const step = this.hash.scratch.size;
                for (let i = 0; i < length; i += step) {
                    const n = Math.min(length - i, step);
                    this.hash.read(position, saddr, n);
                    position += BigInt(n);
                    target.set(this.wasmModule.HEAPU8.subarray(saddr, saddr + n), i);
                }
                return length;
            },
            read: (bytes) => {
                bytes = Math.min(bytes, Number(maxHashBytes - position));
                const out = this.alloc(bytes);
                const saddr = hash.scratch.grow(bytes);
                const step = hash.scratch.size;
                for (let i = 0; i < bytes; i += step) {
                    const n = Math.min(bytes - i, step);
                    hash.read(position, saddr, n);
                    position += BigInt(n);
                    out.set(this.wasmModule.HEAPU8.subarray(saddr, saddr + n), i);
                }
                return out;
            },
            view: (bytes) => {
                bytes = Math.min(bytes, Number(maxHashBytes - position));
                const saddr = hash.scratch.grow(bytes);
                hash.read(position, saddr, Math.min(bytes, hash.scratch.size));
                position += BigInt(bytes);
                return this.wasmModule.HEAPU8.subarray(saddr, saddr + bytes);
            },
            *[Symbol.iterator]() {
                const stepSize = 1024n;
                while (position < maxHashBytes) {
                    const bytes = maxHashBytes - stepSize < stepSize ? maxHashBytes - stepSize : stepSize;
                    yield reader.view(Number(bytes));
                }
            },
        };
        return reader;
    }
    dispose() {
        this.hash.dispose();
    }
}
