// @c4312/blake3-native/esm/hash-instance.js
import { inputToArray } from '@c4312/blake3-internal';
import { Transform } from 'stream';
//@ts-ignore
import * as native from '../build/Release/blake3';
export const createHash = () => new NodeHash(native.createHash());
export const createKeyed = (key) => new NodeHash(native.createKeyed(inputToArray(key)));
export const createDeriveKey = (material) => new NodeHash(native.createDeriveKey(inputToArray(material)));
/**
 * @inheritdoc
 */
export class NodeHash extends Transform {
    constructor(hash) {
        super();
        this.hash = hash;
    }
    /**
     * @reader
     */
    reader() {
        return this.hash.reader();
    }
    /**
     * @inheritdoc
     */
    update(data, encoding) {
        this.hash.update(encoding && typeof data === 'string' ? Buffer.from(data, encoding) : inputToArray(data));
        return this;
    }
    /**
     * @inheritdoc
     */
    dispose() {
        // no-op
    }
    digest(encoding, options) {
        let resolvedOpts;
        let resolvedEnc;
        if (encoding && typeof encoding === 'object') {
            resolvedOpts = encoding;
            resolvedEnc = undefined;
        }
        else {
            resolvedOpts = options;
            resolvedEnc = encoding;
        }
        const result = this.hash.digest(resolvedOpts?.length);
        return resolvedEnc ? result.toString(resolvedEnc) : result;
    }
    /**
     * @inheritdoc
     * @hidden
     */
    _transform(chunk, encoding, callback) {
        this.update(chunk, encoding);
        callback();
    }
    /**
     * @inheritdoc
     * @hidden
     */
    _flush(callback) {
        callback(null, this.digest());
    }
}
