// @c4312/blake3-native/dist/hash-instance.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeHash = exports.createDeriveKey = exports.createKeyed = exports.createHash = void 0;
const blake3_internal_1 = require("@c4312/blake3-internal");
const stream_1 = require("stream");
//@ts-ignore
const native = __importStar(require("../build/Release/blake3"));
const createHash = () => new NodeHash(native.createHash());
exports.createHash = createHash;
const createKeyed = (key) => new NodeHash(native.createKeyed((0, blake3_internal_1.inputToArray)(key)));
exports.createKeyed = createKeyed;
const createDeriveKey = (material) => new NodeHash(native.createDeriveKey((0, blake3_internal_1.inputToArray)(material)));
exports.createDeriveKey = createDeriveKey;
/**
 * @inheritdoc
 */
class NodeHash extends stream_1.Transform {
    constructor(hash) {
        super();
        this.hash = hash;
    }
    /**
     * @reader
     */
    reader() {
        return this.hash.reader();
    }
    /**
     * @inheritdoc
     */
    update(data, encoding) {
        this.hash.update(encoding && typeof data === 'string' ? Buffer.from(data, encoding) : (0, blake3_internal_1.inputToArray)(data));
        return this;
    }
    /**
     * @inheritdoc
     */
    dispose() {
        // no-op
    }
    digest(encoding, options) {
        let resolvedOpts;
        let resolvedEnc;
        if (encoding && typeof encoding === 'object') {
            resolvedOpts = encoding;
            resolvedEnc = undefined;
        }
        else {
            resolvedOpts = options;
            resolvedEnc = encoding;
        }
        const result = this.hash.digest(resolvedOpts?.length);
        return resolvedEnc ? result.toString(resolvedEnc) : result;
    }
    /**
     * @inheritdoc
     * @hidden
     */
    _transform(chunk, encoding, callback) {
        this.update(chunk, encoding);
        callback();
    }
    /**
     * @inheritdoc
     * @hidden
     */
    _flush(callback) {
        callback(null, this.digest());
    }
}
exports.NodeHash = NodeHash;
